src/lib.rs

```diff
@@ -33,34 +34,37 @@ impl Into<ffi::ftdi_interface> for Interface {
     }
 }
 
+pub type Result<T> = result::Result<T, error::Error>;
 
 pub struct Context {
-    native: ffi::ftdi_context,
+    native: *mut ffi::ftdi_context,
 }
 
 impl Context {
-    pub fn new() -> Context {
-        let mut context = Context { native: Default::default() };
-        let result = unsafe { ffi::ftdi_init(&mut context.native) };
+    fn mk_error(&self, res: raw::c_int) -> error::LibFtdiReturn {
+        error::LibFtdiReturn::new(res, &self)
+    }
+
+    pub fn new() -> Result<Context> {
+        let ctx = unsafe { ffi::ftdi_new() };
+
         // Can be non-zero on either OOM or libusb_init failure
-        assert!(result == 0);
-        context
+        if ctx.is_null() {
+            Err(Error::MallocFailure)
+        } else {
+            Ok(Context { native: ctx })
+        }
     }
 
     /// Do not call after opening the USB device
-    pub fn set_interface(&mut self, interface: Interface) -> io::Result<()> {
-        let result = unsafe { ffi::ftdi_set_interface(&mut self.native, interface.into()) };
-        match result {
-            0 => Ok(()),
-            -1 => Err(io::Error::new(ErrorKind::InvalidInput, "unknown interface")),
-            -2 => Err(io::Error::new(ErrorKind::NotFound, "device not found")),
-            -3 => Err(io::Error::new(ErrorKind::Other, "device already opened")),
-            _ => Err(io::Error::new(ErrorKind::Other, "unknown set latency error")),
-        }
+    pub fn set_interface(&mut self, interface: Interface) -> Result<()> {
+        let result = unsafe { ffi::ftdi_set_interface(self.native, interface.into()) };
+
+        self.mk_error(result).into()
```

Comment:
Is `mk_error` ever going to be called without a subsequent `into`? If so I think that it would make more sense to merge these functions.

Reply:
>Is `mk_error` ever going to be called without a subsequent into?

No. And I just realized `into()` should work on shared references: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=65a271479121487431e666b91af1260b

>If not I think that it would make more sense to merge these functions.

I was asking myself "how do I pass in the return type?". Well, I can add private fields to `Context` to make the `Into` impl work.

Reply:
I was more thinking just remove the `Into` impl entirely and have `mk_error` return a `Result<u32, Error>` directly